### PR TITLE
VZ-6511 VAO Metrics Exporter Metrics End to End tests

### PR DESCRIPTION
**Pull request by Raaghavcodes**
Includes end to end tests for Reconcile successful, fail, and duration for the appconfig, cohworkload, helidonworkload, and ingresstrait controllers in VAO.
Files changed for end to end test: networkpolicy.yaml & metrics_test.go

### CODE REVIEW

File: application-operator/controllers/appconfig/appconfig_controller_test.go [hunk 1]
```diff
@@ -729,3 +744,20 @@ func TestReconcileKubeSystem(t *testing.T) {
 	assert.Nil(err)
 	assert.True(result.IsZero())
 }
+
+func TestReconcileFailed(t *testing.T) {
+	metricsexporter.RequiredInitialization()
+	assert := assert.New(t)
+	_ = oamcore.AddToScheme(k8scheme.Scheme)
+	clientBuilder := fake.NewClientBuilder().WithScheme(k8scheme.Scheme).Build()
```
Comment (davisgomes) [inline application-operator/controllers/appconfig/appconfig_controller_test.go:752]: You aren't using the scheme for your test here. Is it required to add oamcore to your scheme or can you get rid of that line?

File: application-operator/controllers/appconfig/appconfig_controller.go [hunk 2]
```diff
@@ -58,6 +56,27 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 	// We do not want any resource to get reconciled if it is in namespace kube-system
 	// This is due to a bug found in OKE, it should not affect functionality of any vz operators
 	// If this is the case then return success
+	zapLogForMetrics := zap.S().With(controllerName)
```
Comment (anders-swanson) [inline application-operator/controllers/appconfig/appconfig_controller.go:59]: looks like these four controllers do the same thing every time they reconcile: get 2 counters and a duration metric, then defer the duration. Did you consider making this generic to reduce the duplication?

File: application-operator/controllers/helidonworkload/helidonworkload_controller.go [hunk 3]
```diff
@@ -73,6 +74,27 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
+		return reconcile.Result{}, err
+	}
+
+	reconcileDurationMetricObject, err := metricsexporter.GetDurationMetric(metricsexporter.HelidonReconcileDuration)
+	if err != nil {
+		zapLogForMetrics.Error(err)
+		return reconcile.Result{}, err
+	}
+	reconcileDurationMetricObject.TimerStart()
+	defer reconcileDurationMetricObject.TimerStop()
```
Comment (davisgomes) [inline application-operator/controllers/helidonworkload/helidonworkload_controller.go:96]: I see this block of code repeated a few times. Is it possible to extrapolate this into a shared function that takes in the metric types and returns the metrics objects?

File: application-operator/metricsexporter/metricsexporter.go [hunk 4]
```diff
@@ -4,21 +4,65 @@
...
-			zap.S().Errorf("Failed to start metrics server for verrazzano-application-operator: %v", err)
-		}
-	}, time.Second*3, wait.NeverStop)
+var (
+	MetricsExp           = metricsExporter{}
+	DefaultLabelFunction func(index int64) string
+	TestDelegate         = metricsDelegate{}
+)
+
+type metricsExporter struct {
```
Comment (anders-swanson) [inline application-operator/metricsexporter/metricsexporter.go:17]: Are these structs any different than George's metricsExporter structs? Can they be reused?

Comment (anders-swanson) [reply application-operator/metricsexporter/metricsexporter.go:17]: The only difference would be the actual metrics objects, right?

File: platform-operator/helm_config/charts/verrazzano/templates/networkpolicy.yaml [hunk 5]
```diff
@@ -119,6 +119,13 @@ spec:
     - ports:
         - port: 9443
           protocol: TCP
+  - from:
```
Comment (anders-swanson) [inline platform-operator/helm_config/charts/verrazzano/templates/networkpolicy.yaml:122]: Add port to netpol

File: tests/e2e/metrics/syscomponents/metrics_test.go [hunk 6]
```diff
@@ -353,3 +339,8 @@ func getClusterNameForPromQuery() string {
 	}
 	return ""
 }
+func eventuallyMetricsContainLabels(metricName string, kv map[string]string) {
```
Comment (anders-swanson) [inline tests/e2e/metrics/syscomponents/metrics_test.go:342]: This function is also duplicated? I think I saw similar in George's or Zaid's PR?

File: application-operator/controllers/ingresstrait/ingresstrait_controller_test.go [hunk 7]
```diff
@@ -4184,3 +4235,19 @@ func createReconcilerWithFake(initObjs ...client.Object) Reconciler {
 	reconciler := newIngressTraitReconciler(cli)
 	return reconciler
 }
+func TestReconcileFailed(t *testing.T) {
+	metricsexporter.RequiredInitialization()
+	assert := asserts.New(t)
+	_ = vzapi.AddToScheme(k8scheme.Scheme)
+	cli := fake.NewClientBuilder().WithScheme(k8scheme.Scheme).Build()
```
Comment (davisgomes) [inline application-operator/controllers/ingresstrait/ingresstrait_controller_test.go:4242]: Same scheme comment. Please check this for all the tests

File: application-operator/metricsexporter/metricsexporter_test.go [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+var logForTest = zap.S()
+
+func TestCollectReconcileMetrics(t *testing.T) {
+	RequiredInitialization()
+	assert := asserts.New(t)
+	test := struct {
+		name string
+	}{
+		name: "Test that reoncile counter is incremented by one when function is Successful & reconcile counter is incremented by one when function is Failed",
```
Comment (davisgomes) [inline application-operator/metricsexporter/metricsexporter_test.go:23]: Can you implement a positive and negative test here? One where the function fails and one when it succeeds.

File: tests/e2e/metrics/syscomponents/metrics_test.go (hunk 6 above)
Comment (davisgomes) [reply tests/e2e/metrics/syscomponents/metrics_test.go:342]: I think so. Please pick up George's function when you rebase with master

File: application-operator/controllers/appconfig/appconfig_controller_test.go (hunk 1 above)
Comment (Raaghavcodes) [reply application-operator/controllers/appconfig/appconfig_controller_test.go:752]: Thanks for catching that. You are right I can omit that line.

File: application-operator/metricsexporter/metricsexporter_test.go (hunk 8 above)
Comment (Raaghavcodes) [reply application-operator/metricsexporter/metricsexporter_test.go]: There is a fail test implemented it is below the successful test and above the duration test

**Review by mbenmclean — APPROVED** [review]

**Pull request merged**